### DISCUSSION TITLE
Trying to display Price with ',' decimal Seperation

### DISCUSSION THREAD

**Discussion by nordic-digit** [filamentphp/filament#6678] [General]
Hello,

I am trying to pass the price using the TextInput->mask method with the Repeater Component via a HasMany Relationship. Everything works fine when saving, but when I open the edit mask, the price is always displayed incorrectly. I am trying to use the German notation (, for decimal separation and . for thousand separation). Here is the following code:

```php
Forms\Components\TextInput::make('price.price_c2h')
    ->columnSpan(6)
    ->mask(fn ($mask) => $mask->patternBlocks([
        'money' => fn (Mask $mask) => $mask
            ->numeric()
            ->decimalSeparator(',')
            ->mapToDecimalSeparator(['.'])
            ->decimalPlaces(2)
            ->padFractionalZeros()
    ])
        ->pattern('money')
    )
    ->suffix('€')
```

Saved in the database:
{"price_c2h":"234.34"}

Output in the form after reloading the mask:
23434,00

I would appreciate any help or guidance on how to resolve this issue. Thank you.

**Comment by pboivin** [comment]
Hi @nordic-digit, I'm not sure if this is related but I had a similar issue when I started using [Laravel Money](https://github.com/cknow/laravel-money). Do you have any kind of special casts on this field on your model?

**Reply by nordic-digit** [reply]
I just cast the price json to an array

**Comment by nordic-digit** [comment]
I have found out that there is a general problem with the output when setting the decimal separator to a comma. As a test, I replaced the JSON with a decimal field and the problem is the same. It is stored correctly in the database, but the output is incorrect.

**Comment by nordic-digit** [comment]
nothing? Guess i'll open an issue if noone knows a workaround

**Reply by pboivin** [reply]
This is strange... I'm not able to reproduce this issue. Here's an example that works for me :
```php
        return TextInput::make($name)
            ->label($label)
            ->numeric()
            ->mask(
                fn (TextInput\Mask $mask) => $mask
                    ->numeric()
                    ->decimalSeparator(',')
                    ->mapToDecimalSeparator(['.'])
                    ->decimalPlaces(2)
                    ->padFractionalZeros()
            )
            ->minValue(0)
            ->maxValue(5)
            ->suffix(' / 5')
            ->required();

```
 I've also tried with `->patternBlocks()` and `->pattern()`. It seems to work. I'm using a `double` column type in the database.

**Reply by nordic-digit** [reply]
Maybe it's because i am using a Repeater which belongs to a HasMany relationship? I'll try it tomorrow without repeater and relationship.

**Comment by nordic-digit** [comment]
ok either way i still got the problem. one of my coworkers has the same. This is weird. I even copy paste your code but it didn't help

**Reply by andrewdwallo** [reply]
I have the same problem as well. Commas can't be used as the decimal separator. It pads two extra zeros to the end of the value unless the amount of cents is zero. So, for example, 70,00 Euros will work but not 70,25 Euros.

**Comment by pekka** [comment]
I have the same problem in a repeater (serving a JSON field) where I'm showing a text field with the following `mask`

```
 $item->mask(fn (TextInput\Mask $mask) => $mask
      ->patternBlocks([
        'money' => fn (Mask $mask) => $mask
            ->numeric()
            ->thousandsSeparator('.')
            ->decimalSeparator(',')
            ->mapToDecimalSeparator(['.'])
            ->decimalPlaces(2)
            ->padFractionalZeros()
            ->normalizeZeros(false)
          ,
    ])->pattern('money'));
```

The `price` from this JSON 

`[{"type": "2", "price": "222.55", "status": "ACTIVE", "period_end": null, "period_start": null, "rebate_group": "23", "special_offer": null}]`

is rendered thus:

![image](https://github.com/filamentphp/filament/assets/262464/a0edcdcb-9452-488f-96e0-803028a624cb)

**Comment by BillyBouman-2B-IT** [comment]
The following comes close to what is to be expected:

```php
Forms\Components\TextInput::make('amount')
    ->label(\__('common.amount'))
    ->mask(fn (TextInput\Mask $mask) => $mask
        ->patternBlocks([
            'money' => fn (TextInput\Mask $mask) => $mask
                ->numeric()
                ->thousandsSeparator('.')
                ->decimalSeparator(','),
        ])
        ->pattern('€money')
    )
    ->required()
    ->numeric()
    ->afterStateHydrated(function (TextInput $component, $state) {
        $component->state(\str_replace('.', ',', $state));
    }),
```

Only part that isn't working correctly with above code is `123.00` which is being displayed as `123`. All values that contain decimals are displayed as expected.
<img width="555" alt="afbeelding" src="https://github.com/filamentphp/filament/assets/127685984/39bad9e0-3cbd-4462-8ae1-dc8848d0043d">
<img width="542" alt="afbeelding" src="https://github.com/filamentphp/filament/assets/127685984/f9bb2385-43ae-4169-a0b9-b93340abe19a">


Notably is that the column version behaves perfectly:
```php
\Filament\Tables\Columns\TextColumn::make('amount')
    ->label(\__('common.amount'))
    ->searchable()
    ->sortable()
    ->money('EUR'),
```
<img width="134" alt="afbeelding" src="https://github.com/filamentphp/filament/assets/127685984/3b0f9a4a-0158-4a36-8fd6-8438b4e08435">

**Reply by pekka** [reply]
Fantastic, thank you!

**Reply by BillyBouman-2B-IT** [reply]
<img width="542" alt="afbeelding" src="https://github.com/filamentphp/filament/assets/127685984/c44b2372-2230-44eb-bf3e-b65e3f6da5fa">

There is a problem with the code I supplied. Upon editing an existing record and saving without editing it says the value must be a number.

**Reply by BillyBouman-2B-IT** [reply]
The "must be a number" was caused by the following part which converted it to string, hence, must be a number:
```php
->afterStateHydrated(function (TextInput $component, $state) {
    $component->state(\str_replace('.', ',', $state));
}),
``` 

The following works flawlessly at this moment:
<img width="559" alt="afbeelding" src="https://github.com/filamentphp/filament/assets/127685984/d49022a6-9ef6-42bf-833a-4437e93a02a5">
```php
Forms\Components\TextInput::make('amount')
    ->label(\__('common.amount'))
    ->required()
    ->prefix('€')
    ->numeric(),
```

**Reply by pekka** [reply]
Thank you! Will try this out and report back if any other issues arise.
